backend/src/org/commcare/suite/model/OfflineUserRestore.java

```diff
@@ -0,0 +1,182 @@
+            return getInMemoryStream();
+        }
+    }
+
+    private InputStream getInMemoryStream() {
+        try {
+            return new ByteArrayInputStream(restore.getBytes("UTF-8"));
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private InputStream getStreamFromReference() {
+        try {
+            Reference local = ReferenceManager._().DeriveReference(reference);
+            return local.getStream();
+        } catch (IOException | InvalidReferenceException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public String getReference() {
+        return reference;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf)
+            throws IOException, DeserializationException {
+        this.recordId = ExtUtil.readInt(in);
+        this.reference = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.restore = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.username = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        this.password = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+    }
+
+    @Override
+    public void writeExternal(DataOutputStream out) throws IOException {
+        ExtUtil.writeNumeric(out, recordId);
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(reference));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(restore));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(username));
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(password));
+    }
+
+    @Override
+    public void setID(int ID) {
+        recordId = ID;
+    }
+
+    @Override
+    public int getID() {
+        return recordId;
+    }
+
+    private void checkThatRestoreIsValid()
+            throws UnfullfilledRequirementsException, IOException, InvalidStructureException,
+            XmlPullParserException {
+
+        TransactionParserFactory factory = new TransactionParserFactory() {
+            @Override
+            public TransactionParser getParser(KXmlParser parser) {
+                String name = parser.getName();
+                if ("registration".equals(name.toLowerCase())) {
+                    return buildUserParser(parser);
+                }
+                return null;
+            }
+        };
+
+        DataModelPullParser parser = new DataModelPullParser(getRestoreStream(), factory, true, false);
+        parser.parse();
+    }
+
+    private TransactionParser buildUserParser(KXmlParser parser) {
+        return new UserXmlParser(parser) {
+
+            @Override
+            protected void commit(User parsed) throws IOException, UnfullfilledRequirementsException {
+                if (!parsed.getUserType().equals(User.TYPE_DEMO)) {
+                    throw new UnfullfilledRequirementsException(
```

Comment:
Traditionally we use UnfillfilledRequrements as a signal for a configuration resource which is currently invalid but the user could make valid through some action (IE: "You need to update CommCare before proceeding"). I think "InvalidStructureException" is the more appropriate error here, which means "The config file that was requested for processing isn't created correctly, and needs to be changed before it will ever worlk.